compiler/modules/types/string.bal

```diff
@@ -1,65 +1,152 @@
 // Implementation specific to basic type string.
 
 type StringSubtype readonly & record {|
+    CharStringSubtype char;
+    NonCharStringSubtype nonChar;
+|};
+
+// allowed = false with empty `values` means `String:char`
+type CharStringSubtype readonly & record {|
+    boolean allowed;
+    string:Char[] values;
+|};
+
+// allowed = false with empty `values` mean all non char strings
+type NonCharStringSubtype readonly & record {|
     boolean allowed;
     string[] values;
 |};
 
 public function stringConst(string value) returns SemType {
-    StringSubtype st = { allowed: true, values: [value] };
+    CharStringSubtype char;
+    NonCharStringSubtype nonChar;
+    
+    if value.length() == 1 {
+        char = { allowed: true, values: [<string:Char>value] };
+        nonChar = { allowed: true, values: [] };
+    }
+    else {
+        char = { allowed: true, values: [] };
+        nonChar = { allowed: true, values: [value] };
+    }
+    return uniformSubtype(UT_STRING, { char, nonChar });
+}
+
+public function stringChar() returns SemType {
+    StringSubtype st = {
+        char: { allowed: false, values: [] }, 
+        nonChar: { allowed: true, values: [] }
+    };
     return uniformSubtype(UT_STRING, st);
 }
 
+function isAllChar(StringSubtype st) returns boolean {
+    return !st.char.allowed && st.char.values.length() == 0;
+}
+
 function stringSubtypeSingleValue(SubtypeData d) returns string? {
```

Comment:
This function should be handling char and nonChar uniformly. I think it can be written more clearly, so that it is obviously correct.